open-craft/learner-profile-dashboard
# Add support for referencing/embedding specific questions

Cf. [SE-953](https://tasks.opencraft.com/browse/SE-953)

This PR makes it possible to embed specific questions from the LPD into a course. (Up until now, it was only possible to embed full LPDs.) Course authors can target a specific question to embed by providing its ID and type via the "Custom Parameters" setting of the LTI Component that's supposed to embed the question.

**Screenshots**

- Setting "Custom Parameters" to specify target question in Studio: ![studio-question-component](https://user-images.githubusercontent.com/961441/56202968-83fdc600-6044-11e9-9cf3-c11cbf6b299a.png)

- A single question embedded into a course: ![lms-question-component](https://user-images.githubusercontent.com/961441/56202975-86f8b680-6044-11e9-86a5-113e2ca90dab.png)

**Test instructions**

1. Create an account on [HGSE stage](https://hgse-stage.opencraft.hosting/) and enroll into the [demo course](https://hgse-stage.opencraft.hosting/courses/course-v1:edX+DemoX+Demo_Course/about).

1. Navigate to [this unit](https://hgse-stage.opencraft.hosting/courses/course-v1:edX+DemoX+Demo_Course/jump_to/block-v1:edX+DemoX+Demo_Course+type@vertical+block@76a8b14f7fd44521816d278c78e777c5).

    - Verify that the questions embedded there render correctly: They should include a "Submit your answer" button and a submission timestamp next to it (which will initially say "Not yet submitted").

    - Answer each question and click "Submit your answer".

        - Observe that the submission timestamp updates to display the time and date of your most recent submission. Make a note of the submission timestamps for both questions.

        - Observe that the answers you provided remain visible.

1. In a new browser tab, navigate to [this unit](https://hgse-stage.opencraft.hosting/courses/course-v1:edX+DemoX+Demo_Course/jump_to/block-v1:edX+DemoX+Demo_Course+type@vertical+block@45c7cedb4bfe46f4a68c78787151cfb5), which contains the LPD to which the questions from the unit you were working with in the previous step belong.

    - Locate the questions (using "Find in page" functionality of your browser) one by one.

    - For each question, observe that the answer displayed matches the answer you provided in the previous step.

    - For each question, scroll down to the end of the parent section (i.e., to the closest "Submit your answers" button that you can find below the question). Observe that the submission timestamp of the parent section matches the submission timestamp of the question that you noted down earlier. (If both questions belong to the same section, the submission timestamp of the section should match the submission timestamp of the question that you submitted last.)

    - For each question, alter your answers slightly and submit them by clicking the "Submit your answers" button of the parent section(s). Make a note of the submission timestamp(s).

1. Switch back to the browser tab displaying individual questions and refresh the page.

    - Observe that answer data updates to match the most recent answers you provided in the previous step.

    - Observe that the submission timestamps update match the submission timestamp(s) from the parent section(s).

1. Repeat the steps above for the following units containing other question types:

    - [Multiple Choice Questions](https://hgse-stage.opencraft.hosting/courses/course-v1:edX+DemoX+Demo_Course/jump_to/block-v1:edX+DemoX+Demo_Course+type@vertical+block@c05f45fcec50466e87ed86f0687349a5)

    - [Ranking Question](https://hgse-stage.opencraft.hosting/courses/course-v1:edX+DemoX+Demo_Course/jump_to/block-v1:edX+DemoX+Demo_Course+type@vertical+block@4b9156e5a4304a819bc189ec8f69c90e)

    - [Likert Scale Questions](https://hgse-stage.opencraft.hosting/courses/course-v1:edX+DemoX+Demo_Course/jump_to/block-v1:edX+DemoX+Demo_Course+type@vertical+block@4d0e01b1ee5c4f90bc3549d0975eed97)

    *Note that some questions are set up to display answer options in random order, so answers might not look the same at a glance even though they actually match.*

**Reviewers**
